File: apps/api/src/execution/scheduler.ts
import cron, { type ScheduledTask } from "node-cron";
import { reconcilePendingSignals } from "./executionPolicy";
import { monitorLiveTrades } from "./liveTrade";
import { runScalpManagementTick } from "./scalpManager";
import { monitorOpenTrades, runWeeklyJournalReview } from "./paperTrading";
import { runDailyBriefing } from "./dailyBriefing";
import { sendDailyNewsBrief } from "./newsBrief";
import { expireStaleApprovals } from "../telegram/approvals";

function isLiveBroker(): boolean {
  return (process.env.BROKER ?? "paper").trim().toLowerCase() === "exness";
}

/** The fast scalp-management loop is opt-in: it only runs live and only when
 *  explicitly enabled, so the 5-min reconciler stays the default everywhere else. */
function scalpManagerEnabled(): boolean {
  return (process.env.ENABLE_SCALP_MANAGER ?? "false").trim().toLowerCase() === "true";
}

/** The morning Telegram news brief is on by default; set ENABLE_NEWS_BRIEF=false to mute it. */
function newsBriefEnabled(): boolean {
  return (process.env.ENABLE_NEWS_BRIEF ?? "true").trim().toLowerCase() !== "false";
}

const PAPER_CRON = "*/5 * * * *";
const SCALP_CRON = "*/15 * * * * *"; // every 15s — scalps need active management, not the 5-min loop
const WEEKLY_CRON = "0 0 * * 0"; // Sunday 00:00 UTC
const DAILY_CRON = "0 6 * * *"; // 06:00 UTC — morning briefing before the session
const EXPIRY_CRON = "* * * * *"; // every minute — approvals are perishable

let paperTask: ScheduledTask | null = null;
let scalpTask: ScheduledTask | null = null;
let weeklyTask: ScheduledTask | null = null;
let dailyTask: ScheduledTask | null = null;
let expiryTask: ScheduledTask | null = null;
let paperRunning = false;
let scalpRunning = false;
let weeklyRunning = false;
let dailyRunning = false;
let expiryRunning = false;

async function runPaperTick(): Promise<void> {
  if (paperRunning) {
    console.log(`[paperCron] ${new Date().toISOString()} prev_run_in_progress, skipping`);
    return;
  }
  paperRunning = true;
  const startedAt = new Date().toISOString();
  console.log(`[paperCron] ${startedAt} run_start`);
  try {
    const rec = await reconcilePendingSignals();
    console.log(
      `[paperCron] ${new Date().toISOString()} reconcile scanned=${rec.scanned} opened=${rec.opened} awaiting=${rec.awaiting} held=${rec.held} blocked=${rec.blocked}`,
    );
    if (isLiveBroker()) {
      const mon = await monitorLiveTrades();
      console.log(
        `[paperCron] ${new Date().toISOString()} live_monitor inspected=${mon.inspected} closed=${mon.closed} unchanged=${mon.unchanged}`,
      );
    } else {
      const mon = await monitorOpenTrades();
      console.log(
        `[paperCron] ${new Date().toISOString()} monitor inspected=${mon.inspected} closed=${mon.closed} unchanged=${mon.unchanged} no_price=${mon.noPrice}`,
      );
    }
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    console.error(`[paperCron] ${new Date().toISOString()} error="${msg}"`);
  } finally {
    paperRunning = false;
    console.log(`[paperCron] ${new Date().toISOString()} run_end`);
  }
}

async function runScalpTick(): Promise<void> {
  if (scalpRunning) return; // a prior tick (e.g. one that closed + journaled) is still running
  scalpRunning = true;
  try {
    const r = await runScalpManagementTick();
    if (r.closed > 0) {
      console.log(
        `[scalpManager] ${new Date().toISOString()} managed=${r.managed} closed=${r.closed} held=${r.held} gone=${r.gone}`,
      );
    }
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    console.error(`[scalpManager] ${new Date().toISOString()} error="${msg}"`);
  } finally {
    scalpRunning = false;
  }
}

async function runWeeklyTick(): Promise<void> {
  if (weeklyRunning) {
    console.log(`[weeklyReviewCron] ${new Date().toISOString()} prev_run_in_progress, skipping`);
    return;
  }
  weeklyRunning = true;
  const startedAt = new Date().toISOString();
  console.log(`[weeklyReviewCron] ${startedAt} run_start`);
  try {
    const r = await runWeeklyJournalReview();
    console.log(
      `[weeklyReviewCron] ${new Date().toISOString()} status=${r.status} trades=${r.tradeCount ?? 0}${r.reason ? ` reason="${r.reason}"` : ""}`,
    );
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    console.error(`[weeklyReviewCron] ${new Date().toISOString()} error="${msg}"`);
  } finally {
    weeklyRunning = false;
  }
}

async function runDailyTick(notify: boolean = true): Promise<void> {
  if (dailyRunning) {
    console.log(`[dailyBriefingCron] ${new Date().toISOString()} prev_run_in_progress, skipping`);
    return;
  }
  dailyRunning = true;
  try {
    await runDailyBriefing();
    // Push the morning news brief to Telegram on the scheduled run only — not on
    // the startup-once pass, so restarts don't spam the chat.
    if (notify && newsBriefEnabled()) {
      const r = await sendDailyNewsBrief();
      console.log(
        `[newsBrief] ${new Date().toISOString()} sent=${r.sent}${r.reason ? ` reason=${r.reason}` : ""}`,
      );
    }
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    console.error(`[dailyBriefingCron] ${new Date().toISOString()} error="${msg}"`);
  } finally {
    dailyRunning = false;
  }
}

async function runExpiryTick(): Promise<void> {
  if (expiryRunning) return;
  expiryRunning = true;
  try {
    const r = await expireStaleApprovals();
    if (r.expired > 0) {
      console.log(`[approvalExpiry] ${new Date().toISOString()} expired=${r.expired}`);
    }
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    console.error(`[approvalExpiry] ${new Date().toISOString()} error="${msg}"`);
  } finally {
    expiryRunning = false;
  }
}

export function startPaperTradingScheduler(): void {
  if (paperTask) return;
  console.log(`[paperCron] scheduled "${PAPER_CRON}"`);
  paperTask = cron.schedule(PAPER_CRON, () => {
    void runPaperTick();
  });
  if (!expiryTask) {
    console.log(`[approvalExpiry] scheduled "${EXPIRY_CRON}"`);
    expiryTask = cron.schedule(EXPIRY_CRON, () => {
      void runExpiryTick();
    });
  }
  if (!scalpTask && isLiveBroker() && scalpManagerEnabled()) {
    console.log(`[scalpManager] scheduled "${SCALP_CRON}"`);
    scalpTask = cron.schedule(SCALP_CRON, () => {
      void runScalpTick();
    });
  }
}

export function startWeeklyReviewScheduler(): void {
  if (weeklyTask) return;
  console.log(`[weeklyReviewCron] scheduled "${WEEKLY_CRON}"`);
  weeklyTask = cron.schedule(WEEKLY_CRON, () => {
    void runWeeklyTick();
  });
}

export function startDailyBriefingScheduler(): void {
  if (dailyTask) return;
  console.log(`[dailyBriefingCron] scheduled "${DAILY_CRON}"`);
  dailyTask = cron.schedule(DAILY_CRON, () => {
    void runDailyTick();
  });
}

export function stopExecutionSchedulers(): void {
  if (dailyTask) {
    dailyTask.stop();
    dailyTask = null;
    console.log("[dailyBriefingCron] stopped");
  }
  if (paperTask) {
    paperTask.stop();
    paperTask = null;
    console.log("[paperCron] stopped");
  }
  if (scalpTask) {
    scalpTask.stop();
    scalpTask = null;
    console.log("[scalpManager] stopped");
  }
  if (weeklyTask) {
    weeklyTask.stop();
    weeklyTask = null;
    console.log("[weeklyReviewCron] stopped");
  }
  if (expiryTask) {
    expiryTask.stop();
    expiryTask = null;
    console.log("[approvalExpiry] stopped");
  }
}

export async function runPaperTradingOnce(): Promise<void> {
  await runPaperTick();
}

export async function runScalpManagementOnce(): Promise<void> {
  await runScalpTick();
}

export async function runWeeklyReviewOnce(): Promise<void> {
  await runWeeklyTick();
}

export async function runDailyBriefingOnce(): Promise<void> {
  await runDailyTick(false); // recompute the briefing on startup, but don't ping Telegram
}

/** On-demand: build and push the news brief to Telegram now (e.g. a manual route). */
export async function sendNewsBriefOnce(): Promise<{ sent: boolean; reason?: string }> {
  return sendDailyNewsBrief();
}

File: services/backend/app/jobs/scheduler.py
"""Scheduled jobs — port of `apps/api/src/execution/scheduler.ts`.

Same cadences as the Express node-cron setup, with two additions the plan
requires:

* **Single owner.** Jobs only start when `BACKEND_JOB_OWNER=true`, and each tick
  additionally takes a short Redis lock, so running several uvicorn workers can
  never double-execute a tick.
* **No overlap.** An in-flight tick is skipped rather than queued, exactly like
  the `*Running` guards in the TypeScript version.

Every tick opens its own session and swallows its own exceptions: a failing job
must not take down the scheduler or the API.
"""
from __future__ import annotations

import logging
from collections.abc import Awaitable, Callable

from apscheduler.schedulers.asyncio import AsyncIOScheduler
from apscheduler.triggers.cron import CronTrigger
from sqlalchemy.ext.asyncio import AsyncSession

from ..core.settings import get_settings
from ..db.redis_client import try_lock
from ..db.session import session_scope
from .clock import utcnow

log = logging.getLogger("backend.scheduler")

PAPER_CRON = "*/5 * * * *"
SCALP_CRON = "*/15 * * * * *"  # every 15s — scalps need active management
WEEKLY_CRON = "0 0 * * 0"  # Sunday 00:00 UTC
DAILY_CRON = "0 6 * * *"  # 06:00 UTC — morning briefing before the session
EXPIRY_CRON = "* * * * *"  # every minute — approvals are perishable

_scheduler: AsyncIOScheduler | None = None
_running: set[str] = set()


def _cron(expr: str) -> CronTrigger:
    """Build a UTC trigger from a 5- or 6-field cron expression."""
    parts = expr.split()
    if len(parts) == 6:
        second, minute, hour, day, month, dow = parts
    else:
        second = "0"
        minute, hour, day, month, dow = parts
    return CronTrigger(
        second=second,
        minute=minute,
        hour=hour,
        day=day,
        month=month,
        day_of_week=dow,
        timezone="UTC",
    )


async def _guarded(
    name: str, lock_ttl_s: int, body: Callable[[AsyncSession], Awaitable[None]]
) -> None:
    """Run one tick: skip if already in flight, hold the Redis lock, never raise."""
    if name in _running:
        log.info("[%s] %s prev_run_in_progress, skipping", name, utcnow().isoformat())
        return
    _running.add(name)
    try:
        async with try_lock(f"jobs:lock:{name}", lock_ttl_s) as acquired:
            if not acquired:
                log.debug("[%s] another process holds the lock, skipping", name)
                return
            async with session_scope() as session:
                await body(session)
    except Exception as exc:
        log.error('[%s] %s error="%s"', name, utcnow().isoformat(), exc)
    finally:
        _running.discard(name)


# --------------------------------------------------------------------------- #
# Tick bodies
# --------------------------------------------------------------------------- #


async def _paper_body(session: AsyncSession) -> None:
    from ..domain.execution.live_trade import monitor_live_trades
    from ..domain.execution.paper_trading import monitor_open_trades
    from ..domain.execution.policy import reconcile_pending_signals

    rec = await reconcile_pending_signals(session)
    log.info(
        "[paperCron] %s reconcile scanned=%d opened=%d awaiting=%d held=%d blocked=%d",
        utcnow().isoformat(),
        rec.scanned,
        rec.opened,
        rec.awaiting,
        rec.held,
        rec.blocked,
    )
    if get_settings().is_live_broker:
        mon = await monitor_live_trades(session)
        log.info(
            "[paperCron] %s live_monitor inspected=%d closed=%d unchanged=%d",
            utcnow().isoformat(),
            mon.inspected,
            mon.closed,
            mon.unchanged,
        )
    else:
        mon = await monitor_open_trades(session)
        log.info(
            "[paperCron] %s monitor inspected=%d closed=%d unchanged=%d no_price=%d",
            utcnow().isoformat(),
            mon.inspected,
            mon.closed,
            mon.unchanged,
            mon.noPrice,
        )


async def _scalp_body(session: AsyncSession) -> None:
    from ..domain.execution.scalp_manager import run_scalp_management_tick

    result = await run_scalp_management_tick(session)
    if result.closed > 0:
        log.info(
            "[scalpManager] %s managed=%d closed=%d held=%d gone=%d",
            utcnow().isoformat(),
            result.managed,
            result.closed,
            result.held,
            result.gone,
        )


async def _weekly_body(session: AsyncSession) -> None:
    from ..domain.execution.paper_trading import run_weekly_journal_review

    result = await run_weekly_journal_review(session)
    log.info(
        "[weeklyReviewCron] %s status=%s trades=%s%s",
        utcnow().isoformat(),
        result.get("status"),
        result.get("tradeCount", 0),
        f' reason="{result["reason"]}"' if result.get("reason") else "",
    )


async def _daily_body(session: AsyncSession, notify: bool = True) -> None:
    from ..domain.execution.daily_briefing import run_daily_briefing
    from ..domain.execution.data_freshness import send_data_freshness_alert
    from ..domain.execution.news_brief import send_daily_news_brief

    await run_daily_briefing(session)
    # Push the morning news brief on the scheduled run only — not on the
    # startup-once pass, so restarts don't spam the chat.
    if notify and get_settings().enable_news_brief:
        result = await send_daily_news_brief(session)
        log.info(
            "[newsBrief] %s sent=%s%s",
            utcnow().isoformat(),
            result.get("sent"),
            f' reason={result["reason"]}' if result.get("reason") else "",
        )
    # Daily staleness alert: scheduled runs only, so restarts don't spam the chat.
    if notify:
        fresh = await send_data_freshness_alert(session)
        if fresh.get("stale") or fresh.get("sent"):
            log.info(
                "[dataFreshness] %s stale=%s sent=%s%s",
                utcnow().isoformat(),
                fresh.get("stale"),
                fresh.get("sent"),
                f' reason={fresh["reason"]}' if fresh.get("reason") else "",
            )


async def _expiry_body(session: AsyncSession) -> None:
    from ..domain.execution.review_agent import expire_stale_recommendations
    from ..integrations.telegram.approvals import expire_stale_approvals

    approvals = await expire_stale_approvals(session)
    if approvals["expired"] > 0:
        log.info("[approvalExpiry] %s expired=%d", utcnow().isoformat(), approvals["expired"])
    recs = await expire_stale_recommendations(session)
    if recs["expired"] > 0:
        log.info(
            "[approvalExpiry] %s recommendations_expired=%d",
            utcnow().isoformat(),
            recs["expired"],
        )


# --------------------------------------------------------------------------- #
# Lifecycle
# --------------------------------------------------------------------------- #


def _tick(name: str, ttl: int, body: Callable[[AsyncSession], Awaitable[None]]):
    async def runner() -> None:
        await _guarded(name, ttl, body)

    return runner


def start_schedulers() -> None:
    """Start the jobs this process owns. No-op unless `BACKEND_JOB_OWNER=true`."""
    global _scheduler
    cfg = get_settings()

    if not cfg.backend_job_owner:
        log.warning("[scheduler] BACKEND_JOB_OWNER=false — no scheduled jobs on this process")
        return
    if cfg.api_shadow_mode:
        # A shadow instance computes decisions but must never act on them.
        log.warning("[scheduler] API_SHADOW_MODE=true — execution jobs stay stopped")
        return
    if _scheduler is not None:
        return

    _scheduler = AsyncIOScheduler(timezone="UTC")

    if cfg.enable_paper_trading:
        _scheduler.add_job(
            _tick("paperCron", 280, _paper_body), _cron(PAPER_CRON), id="paperCron"
        )
        log.info('[paperCron] scheduled "%s"', PAPER_CRON)
        _scheduler.add_job(
            _tick("approvalExpiry", 55, _expiry_body), _cron(EXPIRY_CRON), id="approvalExpiry"
        )
        log.info('[approvalExpiry] scheduled "%s"', EXPIRY_CRON)
        if cfg.is_live_broker and cfg.enable_scalp_manager:
            _scheduler.add_job(
                _tick("scalpManager", 14, _scalp_body), _cron(SCALP_CRON), id="scalpManager"
            )
            log.info('[scalpManager] scheduled "%s"', SCALP_CRON)

    if cfg.enable_weekly_review:
        _scheduler.add_job(
            _tick("weeklyReviewCron", 3600, _weekly_body), _cron(WEEKLY_CRON), id="weeklyReviewCron"
        )
        log.info('[weeklyReviewCron] scheduled "%s"', WEEKLY_CRON)

    if cfg.enable_daily_briefing:
        _scheduler.add_job(
            _tick("dailyBriefingCron", 900, _daily_body), _cron(DAILY_CRON), id="dailyBriefingCron"
        )
        log.info('[dailyBriefingCron] scheduled "%s"', DAILY_CRON)

    _scheduler.start()


def stop_schedulers() -> None:
    """Stop every job. Called on shutdown and before an execution-owner handover."""
    global _scheduler
    if _scheduler is None:
        return
    _scheduler.shutdown(wait=False)
    _scheduler = None
    log.info("[scheduler] stopped")


def scheduler_state() -> dict[str, object]:
    """What this process is running — surfaced by `/health/ready` for observability."""
    cfg = get_settings()
    jobs = [] if _scheduler is None else [j.id for j in _scheduler.get_jobs()]
    return {
        "jobOwner": cfg.backend_job_owner,
        "shadowMode": cfg.api_shadow_mode,
        "running": sorted(jobs),
        "inFlight": sorted(_running),
    }


# --- on-demand run (startup) ---


async def run_daily_briefing_once() -> None:
    """Recompute the briefing on startup, but don't ping Telegram.

    The only on-demand entry point with a caller (`app/main.py`): a restart gets
    a fresh summary without spamming the chat.
    """

    async def body(session: AsyncSession) -> None:
        await _daily_body(session, notify=False)

    await _guarded("dailyBriefingCron", 900, body)
